### PR TITLE
Elasticsearch: Index events and keep them in sync

### LINKED ISSUE #3467: Implement elasticsearch for Indexing and search of events
Implement [elasticsearch](https://github.com/elastic/elasticsearch) so that we can index and search large number of events easily and efficiently. 

Required:
- Adding of event to the elasticsearch index on creation
- Updating of event in the elasticsearch index on edit
- Deleting of event from elasticsearch index when the event is deleted
- Search API to search using elasticsearch
- Experiment on the fuzziness value and choose appropriately. 

This is a parent issue. Sub-issues can be created.

### LINKED ISSUE #4448: Make celery follow flask application factories pattern
http://flask.pocoo.org/docs/0.12/patterns/appfactories/

### PR DESCRIPTION
**Pull request by shubham-padia**
fixes #3467 
fixes #4448 

- Adds insert/update/delete events to their respective redis sets, a scheduler job syncs them via the bulk API at regular intervals of time, currently 60 mins.
- Rebuild entire event index at start of app, which will be at every deployment
- Run a cron job every week to rebuild the entire index from scratch

To know more about how and why this approach has been followed,
visit https://www.elastic.co/blog/found-keeping-elasticsearch-in-sync

To know more about other decisions like weekly rebuilds, visit https://gocardless.com/blog/syncing-postgres-to-elasticsearch-lessons-learned/

### CODE REVIEW

File: app/views/elastic_search.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+es = FlaskElasticsearch()
+
+
+def rebuild_events_elasticsearch():
+    """
+    Re-inserts all eligible events into elasticsearch
+    Also clears event_index and event_delete redis sets
+    :return:
+    """
+    events = Event.query.filter(Event.state == 'published', Event.deleted_at == None)
```
[the remark this answers is not in the document]
Comment (shubham-padia) [reply app/views/elastic_search.py]: To be ignored.

File: app/views/elastic_cron_helpers.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+def cron_rebuild_events_elasticsearch():
+    """
+    Re-inserts all eligible events into elasticsearch
+    Also clears event_index and event_delete redis sets
+    :return:
+    """
+    cur = conn.cursor()
+    cur.execute(
+        "SELECT id, name, description, searchable_location_name, organizer_name, organizer_description FROM events WHERE state = 'published' and deleted_at is NULL ;")
```
[the remark this answers is not in the document]
Comment (shubham-padia) [reply app/views/elastic_cron_helpers.py]: To be ignored.

File: app/views/elastic_cron_helpers.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    def next(self):
+        if self.current >= self.high:
+            raise StopIteration
+        else:
+            self.current += 1
+            event_id = redis_store.spop(self.set_name)
+            cur = conn.cursor()
+            cur.execute(
+                "SELECT id, name, description, searchable_location_name, organizer_name, organizer_description FROM events WHERE id = %s;",
```
[the remark this answers is not in the document]
Comment (shubham-padia) [reply app/views/elastic_cron_helpers.py]: To be ignored.

File: app/__init__.py [hunk 4]
```diff
@@ -42,6 +40,10 @@
 from app.models.role_invite import RoleInvite
 from app.views.healthcheck import health_check_celery, health_check_db, health_check_migrations, check_migrations
 from app.views.sentry import sentry
+from app.views.elastic_search import es, rebuild_events_elasticsearch
```
[the remark this answers is not in the document]
Comment (shubham-padia) [reply app/__init__.py]: To be ignored.

File: app/__init__.py [hunk 5]
```diff
@@ -42,6 +40,10 @@
...
+from app.views.elastic_cron_helpers import sync_events_elasticsearch, cron_rebuild_events_elasticsearch
```
[the remark this answers is not in the document]
Comment (shubham-padia) [reply app/__init__.py]: To be ignored.

File: app/__init__.py [hunk 6]
```diff
@@ -42,6 +40,10 @@
...
+from app.views.redis_store import redis_store
```
[the remark this answers is not in the document]
Comment (shubham-padia) [reply app/__init__.py]: To be ignored.

**Comment by niranjan94** [comment]
> Adds insert/update/delete events to their respective redis sets, a scheduler job syncs them via the bulk API at regular intervals of time, currently 60 mins.

Why the every 60 mins job ? Couldn't you make celery do it ? It would be much cleaner instead of adding a record manually to redis and having a 60 min cron to handle it. 

> Rebuild entire event index at start of app, which will be at every deployment

That would make the app start insanely slow once there are a lot of record. Instead, start a celery task when the app is starting and let celery take care of it in the background

> Run a cron job every week to rebuild the entire index from scratch

Let this also not be on the app server and instead be handled by celery.

**Comment by shubham-padia** [comment]
> Why the every 60 mins job ? Couldn't you make celery do it ? It would be much cleaner instead of adding a record manually to redis and having a 60 min cron to handle it.

No, celery cannot do this. Celery will add tasks in job queue. Adding individual tasks in job queue defeats the whole purpose of going through this approach.

> That would make the app start insanely slow once there are a lot of record. Instead, start a celery task when the app is starting and let celery take care of it in the background.
Let this also not be on the app server and instead be handled by celery

Can use celery here I think, will check

**Comment by shubham-padia** [comment]
> That would make the app start insanely slow once there are a lot of record. Instead, start a celery task when the app is starting and let celery take care of it in the background.
Let this also not be on the app server and instead be handled by celery

@niranjan94 Made both of the above celery tasks, please review :smile:

**Comment by shubham-padia** [comment]
@niranjan94 Please review :)

**Pull request merged**